### DISCUSSION TITLE
prisma: calling create on model with foreign key gives "'`null` is not assignable to type `undefined`"

### DISCUSSION THREAD

**Discussion by alf5** [TypedProject/ts-express-decorators#2250] [General]
### to recreate
Create a new project with Prisma:
```
❯ npx @tsed/cli init tsed-test
Need to install the following packages:
  @tsed/cli@4.2.7
Ok to proceed? (y) 
npm WARN deprecated @npmcli/move-file@2.0.1: This functionality has been moved to @npmcli/fs
? What is your project name tsed-test
? Choose the target platform: Express.js
? Choose the architecture for your project: Ts.ED
? Choose the convention file styling: Ts.ED
? Check the features needed for your project Database
? Choose a ORM manager Prisma
? Choose the package manager: NPM
↓ Write RC files [SKIPPED]
↓ Write RC files [SKIPPED]
✔ Install plugins
✔ Load plugins
✔ Install plugins dependencies
✔ Generate project files
✔ Generate Prisma schema
✔ Add Ts.ED configuration to Prisma schema
✔ Install dependencies
✔ Generate barrels files
```

Prisma schema:
```prisma
generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

generator tsed {
  provider = "tsed-prisma"
}

model Account {
  /// @TsED.Groups("!creation")
  id            String    @id @default(uuid()) @db.Uuid
  /// @TsED.Email()
  /// @TsED.Description("User email. This email must be unique!")
  email         String    @unique
  /// @TsED.Groups("!creation")
  reseller_id   String?   @db.Uuid
  /// @TsED.Groups("!creation")
  reseller      Reseller? @relation(fields: [reseller_id], references: [id])
}

model Reseller {
  id              String      @id @default(uuid()) @db.Uuid
  /// @TsED.Email()
  /// @TsED.Description("User email. This email must be unique!")
  email           String      @unique
  accounts        Account[]
}
```
executed: `npm run prisma:generate` - all good

executed: `npx @tsed/cli g`
```
❯ npx @tsed/cli g
[10:20:20] ✔ @tsed/cli-plugin-prisma module loaded
? Which type of provider? Controller
? Which name? Account
? Which route? /account
? Which directory? rest
✔ Generate controller file to 'controllers/rest/AccountController.ts'
↓ Update bin/index [SKIPPED]
```

`controllers/AccountController.ts`: - This is pretty much following the [docs](https://tsed.io/tutorials/prisma.html#create-controllers)
```typescript =
import { BodyParams } from "@tsed/common";
import {Controller, Inject} from "@tsed/di";
import { AccountModel, AccountsRepository } from "@tsed/prisma";
import {Groups, Post, Returns, Summary} from "@tsed/schema";

@Controller("/account")
export class AccountController {
  @Inject()
  protected repo: AccountsRepository

  @Post("/")
  @Summary('Creates a new Account')
  @Returns(201, AccountModel)
  async post(@BodyParams() @Groups('creation') data: AccountModel) {
    return this.repo.create({data})
  }
}
```

both VScode and `npm run build` give an error for `this.repo.create({data})` saying that:
```
Types of property 'reseller_id' are incompatible.
 ...
     Type 'null' is not assignable to type 'undefined'
```

Was wondering what I might be doing wrong? or is this a bug? or is there a workaroud for this?

**Comment by Romakita** [comment]
Hello @alf5 

You can try to change the `reseller_id   String?   @db.Uuidˋ by ˋreseller_id   String  @db.Uuid`.

maybe the generated AccounModel gor the reseller fied is incorrect.

if it’s a bug please open issue instead !

See you

**Comment by alf5** [comment]
Hi @Romakita 
It still errors.

I believe this is to do with the generated Model confusing the Prisma CreateArgs type.
The `data` element is an `XOR` of two types whose only difference is one takes the raw DB value (String in this case), the other type takes an Object definition of the Foreign key table - TSED generated Model classes have both defined.

Another Note: TSED generated Model classes have `@Required()` set for properties that have a `default` value defined, this means the validation code on these classes are not appropriate when _creating_, those properties should be `Optional`.
The `@Groups` technique doesn't fit this use-case.

---

I think ideally the TSED code generator should generate seperate classes for create, update and entity (with the appropriate validation decorators).

If you follow what the Prisma generator does it would probably be a major undertaking (??) (see the output of [prisma-zod-generator](https://github.com/omar-dulaimi/prisma-zod-generator) 

Maybe as a "good enough" implementation would be to just support the raw DB values and generate create and update classes without properties that are the "Foreign" objects? (with all the validation decorators)

**Reply by Romakita** [reply]
Ok. Open issue with a small repo example and I’ll a time to fix that.

Note: the Groups is designed to have only one class with all Groups strategy. Generate many class isn’t the purpose of Groups decorator and I won’t change that.

> Another Note: TSED generated Model classes have @Required() set for properties that have a default value defined, this means the validation code on these classes are not appropriate when creating, those properties should be Optional.

this is why you have RequiredGroups decorator and Groups decorator. Please see the doc ;)
